### PR TITLE
Docs: How To Define Portal API

### PR DESCRIPTION
**Pull request by thomasvidas**

### CODE REVIEW

**Comment by carlpoole** [comment]
I think under the `## Implementing the API` section we will need to advise people to add the Capacitor core dependencies to their projects in order to access the Capacitor Plugin classes they need to use.

Eg. for android they need to add it along with the Portals dependency like this

```
dependencies {
    implementation 'io.ionic:portals:0.2.0'
    implementation 'com.capacitorjs:core:3.2.2'
}
```

File: website/docs/how-to/define-api-in-typescript.md [hunk 1]
```diff
@@ -2,3 +2,197 @@
...
+One of the biggest benefits of including Ionic Portals in an application is the ability to easily communicate between web and native code using the [PortalsPlugin](../reference/web/portals-plugin). However, in some more niche cases, creating your own Plugins may be neccessary. By creating a [Capacitor Plugin](https://capacitorjs.com/docs/plugins/creating-plugins), you can create your own API to communicate between web and native code.
+
+For this example, we will create a Plugin called `EchoPlugin` that has a single function: `echo`.
+
+## Creating API definitions
+
+We strongly recommend using TypeScript to create a type defintion file which can be used to define your API. This way, there is a central source of truth for the API across Android and iOS as well as having type defintions for the web code.
+
+```typescript
+export interface IEchoPlugin {
```
Comment (elylucas) [inline website/docs/how-to/define-api-in-typescript.md:18]: not a big deal, but TS nomenclature doesn't normally put "I" in front of their interface names like they do in Java and .Net.

File: website/docs/how-to/define-api-in-typescript.md [hunk 2]
```diff
@@ -2,3 +2,197 @@
...
+   echo(options: { value: string }): Promise<{ value: string }>;
+}
+```
+
+On the Android or iOS side, the `EchoPlugin` class will need to match this type signature.
```
Comment (elylucas) [inline website/docs/how-to/define-api-in-typescript.md:23]: Technically not correct. The swift/java classes have method names that match up, but their parameters and return types aren't the same as the interface. Maybe something like:

"On the Android and iOS side, the `EchoPlugin` class will implement methods that listen for calls made to methods on the `EchoPlugin` interface."

File: website/docs/how-to/define-api-in-typescript.md [hunk 3]
```diff
@@ -2,3 +2,197 @@
...
+public class EchoPlugin: CAPPlugin {
+  @objc func echo(_ call: CAPPluginCall) {
+    let value = call.getString("value");
+    print(value);
+  }
+}
+```
+
+:::info
+In iOS, if you require Objective-C compatibility, you'll also need a file called `EchoPlugin.m` to create Objective-C bindings with helper functions that Capacitor provides. Below is what the `EchoPlugin` Objective-C bindings would look like.
```
Comment (elylucas) [inline website/docs/how-to/define-api-in-typescript.md:70]: nice info section on obj-c compat

**Comment by thomasvidas** [comment]
@carlpoole added 👍

**Review by carlpoole — APPROVED** [review]

**Review by elylucas — COMMENTED** [review]
In the code blocks, we can specify a title like so:
```swift title=AppDelegate.swift```
It adds the title to the top of the code block and helps the user identify where the code goes. I suggest we use the title whenever appropriate.

**Review by elylucas — COMMENTED** [review]
looks good, just a few small clarifications

**Comment by thomasvidas** [comment]
Since this got merged in right as you reviewed this, I'll go ahead a make a new PR and reference your comments there 👍

**Pull request merged**